extensions/resteasy-reactive/quarkus-resteasy-reactive-links/runtime/src/main/java/io/quarkus/resteasy/reactive/links/runtime/hal/HalServerResponseFilter.java

```diff
@@ -0,0 +1,91 @@
+package io.quarkus.resteasy.reactive.links.runtime.hal;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+import javax.ws.rs.container.ContainerRequestContext;
+import javax.ws.rs.container.ContainerResponseContext;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+import org.jboss.resteasy.reactive.common.util.RestMediaType;
+import org.jboss.resteasy.reactive.server.ServerResponseFilter;
+import org.jboss.resteasy.reactive.server.core.CurrentRequestManager;
+
+import io.quarkus.hal.HalCollectionWrapper;
+import io.quarkus.hal.HalEntityWrapper;
+import io.quarkus.hal.HalService;
+import io.quarkus.resteasy.reactive.links.RestLink;
+
+public class HalServerResponseFilter {
+
+    private static final String COLLECTION_NAME = "items";
+
+    private final HalService service;
+
+    @Inject
+    public HalServerResponseFilter(HalService service) {
+        this.service = service;
+    }
+
+    @ServerResponseFilter
+    public void filter(ContainerRequestContext requestContext, ContainerResponseContext responseContext, Throwable t) {
+        if (t == null) {
+            Object entity = responseContext.getEntity();
+            if (isHttpStatusSuccessful(responseContext.getStatusInfo())
+                    && acceptsHalMediaType(requestContext)
+                    && canEntityBeProcessed(entity)) {
+                if (entity instanceof Collection) {
+                    responseContext.setEntity(service.toHalCollectionWrapper((Collection<Object>) entity, COLLECTION_NAME,
+                            findEntityClass()));
+                } else {
+                    responseContext.setEntity(service.toHalWrapper(entity));
+                }
+            }
+        }
+    }
+
+    private boolean canEntityBeProcessed(Object entity) {
+        return entity != null
+                && !(entity instanceof String)
+                && !(entity instanceof HalEntityWrapper || entity instanceof HalCollectionWrapper);
+    }
+
+    private boolean isHttpStatusSuccessful(Response.StatusType statusInfo) {
+        return Response.Status.Family.SUCCESSFUL.equals(statusInfo.getFamily());
+    }
+
+    private boolean acceptsHalMediaType(ContainerRequestContext requestContext) {
+        List<String> acceptMediaType = requestContext.getAcceptableMediaTypes().stream().map(MediaType::toString).collect(
+                Collectors.toList());
+        return acceptMediaType.contains(RestMediaType.APPLICATION_HAL_JSON);
+    }
+
+    private Class<?> findEntityClass() {
```

Comment:
We can leave addressing this for a following PR, but in RESTEasy Reactive we generally try to avoid performing reflection as much as possible as it has a negative effect on both the execution speed and memory consumption. This data is generally available at build time, so we try to leverage it as much as possible.

Reply:
I've removed the second part of this method. For the first part, we can do it in the next iteration.

Reply:
So the second part was not needed?

Reply:
No, it wasn't needed at all.